src/api-v1.js

```diff
@@ -0,0 +1,280 @@
+  //deferAuth: false,
+  //scopes: [],
+  title: "Redirect to backing cache", 
+  description: [
+    'Redirect to the copy of :url in :region.  If there is',
+    'no copy of the file in that region, submit a request to',
+    'backend process to copy into that region and wait to respond',
+    'here until that happens',
+    '',
+    'NOTE: URL parameter must be URL Encoded!',
+    '',
+    'NOTE: If using this with an api-reference consuming client',
+    'you will need to pass error as an empty string!',
+  ].join('\n'),
+}, async function (req, res) {
+  let url = req.params.url;
+  let region = req.params.region;
+  let service = req.params.service;
+  let error = req.params.error;
+  if (error) {
+    return res.status(400).json({
+      msg: 'URL Must be URL encoded!',
+    });
+  }
+
+  let logthingy = `${url} in ${service}/${region}`;
+  debug(`Attempting to redirect to ${logthingy}`);
+
+  if (service.toLowerCase() === 's3') {
+    if (this.s3backends[region]) {
+      let backend = this.s3backends[region];
+      try {
+        await backend.validateInputURL(url);
+      } catch (err) {
+        debug(err);
+        debug(err.stack);
+        // Intentionally vague because we don't want to reveal
+        // our configuration too much
+        let msg = 'Input URL failed validation for unknown reason';
+        switch (err.code) {
+          case 'HTTPError':
+            msg = 'HTTP Error while trying to resolve redirects';
+            break;
+          case 'DoesNotMatchPatterns':
+            msg = 'Input URL does not match whitelist';
+            break;
+          case 'InsecureURL':
+            msg = 'Refusing to follow a non-SSL redirect';
+            break;
+        }
+        return res.status(400).json({
+          msg: msg,
+        });
+      }
+
+      let maxWaitForCachedCopy = 90 * 1000;
+
+      let startTime = new Date();
+      let x = 0;
+      while ((new Date() - startTime) < maxWaitForCachedCopy) {
+        debug(`Check ${x++} of ${url}`);
+        let result = await backend.getBackendUrl(url);
+        if (result.status === 'present') {
+          debug(`${logthingy} is present`);
+          res.status(302);
+          res.location(result.url);
+          // Instead of just returning result object, we want to return only
+          // known properties.  This is to avoid possible leakage
+          let datapoint = {
+            url: url,
+            service: service,
+            region: region,
+          }
+          debug(`Found ${url}`);
+          return res.json({
+            status: result.status,
+            url: result.url,
+          });
+        }
+        // When we have Influx set up, let's submit this datapoint
+        // to a series called 'Cloud Mirror Cache Hits'
+        await delayer(1000);
+      }
+      // If we get here, we're doing the fallback of redirecting
+      // to the original URL
+      debug(`Redirecting to uncached copy because it took too long ${url}`);
```

Comment:
nit... But generally don't use string interpolation with `debug`. Using `debug("bla bla %s bla bla", url);` is much better as `debug` renders to `function() {}` if it's deactivated... So you get better performance from not using string interpolation...